firebase-app-distribution/src/main/java/com/google/firebase/appdistribution/CheckForUpdateClient.java

```diff
@@ -0,0 +1,194 @@
+import com.google.firebase.installations.FirebaseInstallationsApi;
+import com.google.firebase.installations.InstallationTokenResult;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+
+public class CheckForUpdateClient {
+  private static final int UPDATE_THREAD_POOL_SIZE = 4;
+
+  private final FirebaseApp firebaseApp;
+  private FirebaseAppDistributionTesterApiClient firebaseAppDistributionTesterApiClient;
+  private final FirebaseInstallationsApi firebaseInstallationsApi;
+
+  private TaskCompletionSource<AppDistributionRelease> checkForUpdateTaskCompletionSource = null;
+  private CancellationTokenSource checkForUpdateCancellationSource;
+  private final Executor checkForUpdateExecutor;
+
+  CheckForUpdateClient(
+      @NonNull FirebaseApp firebaseApp,
+      @NonNull FirebaseAppDistributionTesterApiClient firebaseAppDistributionTesterApiClient,
+      @NonNull FirebaseInstallationsApi firebaseInstallationsApi) {
+    this.firebaseApp = firebaseApp;
+    this.firebaseAppDistributionTesterApiClient = firebaseAppDistributionTesterApiClient;
+    this.firebaseInstallationsApi = firebaseInstallationsApi;
+    // TODO: verify if this is best way to use executorservice here
+    this.checkForUpdateExecutor = Executors.newFixedThreadPool(UPDATE_THREAD_POOL_SIZE);
+  }
+
+  @NonNull
+  public Task<AppDistributionRelease> checkForUpdate() {
+
+    if (checkForUpdateTaskCompletionSource != null
+        && !checkForUpdateTaskCompletionSource.getTask().isComplete()) {
+      checkForUpdateCancellationSource.cancel();
+    }
+
+    checkForUpdateCancellationSource = new CancellationTokenSource();
+    checkForUpdateTaskCompletionSource =
+        new TaskCompletionSource<>(checkForUpdateCancellationSource.getToken());
+
+    Task<String> installationIdTask = firebaseInstallationsApi.getId();
+    // forceRefresh is false to get locally cached token if available
+    Task<InstallationTokenResult> installationAuthTokenTask =
+        firebaseInstallationsApi.getToken(false);
+
+    Tasks.whenAllSuccess(installationIdTask, installationAuthTokenTask)
+        .addOnSuccessListener(
+            tasks -> {
+              String fid = installationIdTask.getResult();
+              InstallationTokenResult installationTokenResult =
+                  installationAuthTokenTask.getResult();
+              checkForUpdateExecutor.execute(
+                  () -> {
+                    try {
+                      AppDistributionReleaseInternal latestRelease =
+                          getLatestReleaseFromClient(
+                              fid,
+                              firebaseApp.getOptions().getApplicationId(),
+                              firebaseApp.getOptions().getApiKey(),
+                              installationTokenResult.getToken());
+                      updateOnUiThread(
+                          () -> {
+                            // TODO: add latest release to storage
```

Comment:
Should this be moved up on the build plan? I don't think updateApp will work until then since cachedLatestRelease isn't set

Reply:
Good call. It won't work. Moving persistence up in the build plan is one option. Let's think about others before submitting this change